types/int_test.go

```diff
@@ -109,3 +110,211 @@ func TestUint(t *testing.T) {
+		testArithInt(t, n2, n1/n2, ni1.Div, ni1.DivRaw)
+		require.Equal(t, func(i1 int64, i2 int64) int64 {
+			if i1 < i2 {
+				return i1
+			}
+			return i2
+		}(n1, n2), MinInt(ni1, ni2).Int64())
+		require.Equal(t, -n1, ni1.Neg().Int64())
+
+		require.Equal(t, n1 == n2, ni1.Equal(ni2))
+		require.Equal(t, n1 > n2, ni1.GT(ni2))
+		require.Equal(t, n1 < n2, ni1.LT(ni2))
+	}
+}
+
+func testArithUint(t *testing.T, n2 uint64, res uint64, fn func(Uint) Uint, fnraw func(uint64) Uint) {
+	require.Equal(t, res, fn(NewUint(n2)).Uint64())
+	require.Equal(t, res, fnraw(n2).Uint64())
+}
+
+func TestConsistencyUint(t *testing.T) {
+	for i := 0; i < 20; i++ {
+
+		// Test identity: uint64 -> Uint/big.Int/string -> uint64
+		n := uint64(rand.Uint32())
+		ni := NewUint(n)
+
+		require.Equal(t, n, ni.Uint64())
+		require.Equal(t, n, ni.BigInt().Uint64())
+		nifromstr, ok := NewUintFromString(strconv.FormatUint(n, 10))
+		require.True(t, ok)
+		require.Equal(t, n, nifromstr.Uint64())
+		require.Equal(t, n, NewUintFromBigInt(big.NewInt(int64(n))).Uint64())
+		require.Equal(t, n, NewUintWithDecimal(n, 0).Uint64())
+
+		// Test purity
+		require.Equal(t, NewUint(n), ni)
+
+		// Test identity: Uint -> uint64/big.Int/string -> Uint
+		ni = NewUint(n)
+
+		require.Equal(t, ni, NewUint(ni.Uint64()))
+		require.Equal(t, ni, NewUintFromBigInt(ni.BigInt()))
+		nifromstr, ok = NewUintFromString(ni.String())
+		require.True(t, ok)
+		require.Equal(t, ni, nifromstr)
+		require.Equal(t, ni, NewUint(ni.BigInt().Uint64()))
+		require.Equal(t, ni, NewUintWithDecimal(n, 0))
+
+		// Test purity
+		require.Equal(t, NewUint(n), ni)
+
+		// Test homomorphicity
+		n1 := uint64(rand.Uint32())
+		ni1 := NewUint(n1)
+		n2 := uint64(rand.Uint32())
+		ni2 := NewUint(n2)
+
+		testArithUint(t, n2, n1+n2, ni1.Add, ni1.AddRaw)
+		if n1 > n2 {
+			testArithUint(t, n2, n1-n2, ni1.Sub, ni1.SubRaw)
+		}
+		testArithUint(t, n2, n1*n2, ni1.Mul, ni1.MulRaw)
+		testArithUint(t, n2, n1/n2, ni1.Div, ni1.DivRaw)
+		require.Equal(t, func(i1 uint64, i2 uint64) uint64 {
+			if i1 < i2 {
+				return i1
+			}
+			return i2
+		}(n1, n2), MinUint(ni1, ni2).Uint64())
+
+		require.Equal(t, n1 == n2, ni1.Equal(ni2))
+		require.Equal(t, n1 > n2, ni1.GT(ni2))
+		require.Equal(t, n1 < n2, ni1.LT(ni2))
+	}
+}
+
+func TestImmutability(t *testing.T) {
+	for i := 0; i < 20; i++ {
+		n := rand.Int63()
+		ni := NewInt(n)
+		z := ZeroInt()
+		o := OneInt()
+
+		ni.Add(z)
+		ni.AddRaw(0)
+		ni.Sub(z)
+		ni.SubRaw(0)
+		ni.Mul(o)
+		ni.MulRaw(1)
+		ni.Div(o)
+		ni.DivRaw(1)
+		ni.Neg()
+
+		require.Equal(t, n, ni.Int64())
+		require.Equal(t, NewInt(n), ni)
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Uint64()
+		ni := NewUint(n)
+		z := ZeroUint()
+		o := OneUint()
+
+		ni.Add(z)
+		ni.AddRaw(0)
+		ni.Sub(z)
+		ni.SubRaw(0)
+		ni.Mul(o)
+		ni.MulRaw(1)
+		ni.Div(o)
+		ni.DivRaw(1)
+
+		require.Equal(t, n, ni.Uint64())
+		require.Equal(t, NewUint(n), ni)
+	}
+
+}
+
+func TestEncoding(t *testing.T) {
```

Comment:
Can this be made table driven or have a test_vectors file. Its really important that there be a constant table for this, as we need regression tests here.